### PR TITLE
Only restrict CMake version in Umpire when `examples` and `rocm` are enabled

### PR DESCRIPTION
**Pull request by msimberg**
https://github.com/spack/spack/pull/27893 restricted CMake to < 3.21 because of https://github.com/spack/spack/issues/27873. However, it looks like that restriction is unnecessarily strict since the build error happens in examples. This changes the constraint to apply only `when="+examples +rocm"` instead of `when="+rocm"`.

### CODE REVIEW

File: var/spack/repos/builtin/packages/umpire/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/umpire/package.py]: I believe this is [an instance of](https://cmake.org/cmake/help/latest/release/3.21.html#id3)

> The AMD ROCm Platform hipcc compiler was identifed by CMake 3.21.0 through 3.21.2 as a distinct compiler with id ROCMClang. This has been removed because it caused regressions.

So it's probably fine to drop the `depends_on` by now or change it into

```suggestion
    conflicts("^cmake@3.21.0:3.21.2", when="+rocm")
```

Comment (msimberg) [reply var/spack/repos/builtin/packages/umpire/package.py]: I think this is different. At least I still get the error in the examples reported by #27873 with CMake 3.23.2.

However, that restriction you're talking about may also be necessary...

**Comment by davidbeckingsale** [comment]
I think adding `entries.append(cmake_cache_string("CMAKE_HIP_ARCHITECTURES", "{0}".format(arch_str)))` on line ~192 might be a solution - can you give that a try @msimberg ?

**Comment by msimberg** [comment]
> I think adding `entries.append(cmake_cache_string("CMAKE_HIP_ARCHITECTURES", "{0}".format(arch_str)))` on line ~192 might be a solution - can you give that a try @msimberg ?

This seems to work for the `HIP_ARCHITECTURES` issues, thanks! However, only for older versions... I tested with 6.0.0 and 4.1.2 and they work. With 2022.03.1 I get this while configuring:
```
-- Configuring done
CMake Error in CMakeLists.txt:
  Target "blt_hip" INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "/tmp/misimber/spack-stage/spack-stage-umpire-2022.03.1-ukyc6ndexz7vvjxvrgo5yjh3y72w3od3/spack-src/HIP_CLANG_INCLUDE_PATH-NOTFOUND/.."

  which is prefixed in the source directory.


CMake Error in CMakeLists.txt:
  Target "blt_hip" INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "/tmp/misimber/spack-stage/spack-stage-umpire-2022.03.1-ukyc6ndexz7vvjxvrgo5yjh3y72w3od3/spack-src/HIP_CLANG_INCLUDE_PATH-NOTFOUND/.."

  which is prefixed in the source directory.


-- Generating done
```
This is with HIP 5.1.3, CMake 3.23.2. The full spec is:

<details>

```
Input spec
--------------------------------
umpire%gcc@11.2.0+rocm

Concretized
--------------------------------
umpire@2022.03.1%gcc@11.2.0+c~cuda+device_alloc~deviceconst+examples~fortran~ipo~numa~openmp+rocm+shared amdgpu_target=gfx90a build_type=RelWithDebInfo tests=none arch=cray-sles15-zen2
    ^blt@0.5.1%gcc@11.2.0 arch=cray-sles15-zen2
        ^cmake@3.23.2%gcc@11.2.0~doc+ncurses+ownlibs~qt build_type=Release arch=cray-sles15-zen2
            ^ncurses@6.2%gcc@11.2.0~symlinks+termlib abi=none arch=cray-sles15-zen2
                ^pkgconf@1.8.0%gcc@11.2.0 arch=cray-sles15-zen2
            ^openssl@1.1.1q%gcc@11.2.0~docs~shared certs=mozilla patches=3fdcf2d arch=cray-sles15-zen2
                ^ca-certificates-mozilla@2022-07-19%gcc@11.2.0 arch=cray-sles15-zen2
                ^perl@5.34.1%gcc@11.2.0+cpanm+shared+threads arch=cray-sles15-zen2
                    ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl patches=b231fcc arch=cray-sles15-zen2
                    ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared arch=cray-sles15-zen2
                        ^diffutils@3.8%gcc@11.2.0 arch=cray-sles15-zen2
                            ^libiconv@1.16%gcc@11.2.0 libs=shared,static arch=cray-sles15-zen2
                    ^gdbm@1.19%gcc@11.2.0 arch=cray-sles15-zen2
                        ^readline@8.1.2%gcc@11.2.0 arch=cray-sles15-zen2
                    ^zlib@1.2.12%gcc@11.2.0+optimize+pic+shared patches=0d38234 arch=cray-sles15-zen2
    ^camp@2022.03.0%gcc@11.2.0~cuda~ipo+rocm~tests amdgpu_target=gfx90a build_type=RelWithDebInfo arch=cray-sles15-zen2
        ^hip@5.1.3%gcc@11.2.0~ipo build_type=Release patches=3b25db8 arch=cray-sles15-zen2
            ^comgr@5.1.3%gcc@11.2.0~ipo build_type=Release arch=cray-sles15-zen2
                ^llvm-amdgpu@5.1.3%gcc@11.2.0~ipo~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_type=Release arch=cray-sles15-zen2
                    ^python@3.9.13%gcc@11.2.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,ebdca64,f2fd060 arch=cray-sles15-zen2
                        ^expat@2.4.8%gcc@11.2.0+libbsd arch=cray-sles15-zen2
                            ^libbsd@0.11.5%gcc@11.2.0 arch=cray-sles15-zen2
                                ^libmd@1.0.4%gcc@11.2.0 arch=cray-sles15-zen2
                        ^gettext@0.21%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=cray-sles15-zen2
                            ^libxml2@2.9.13%gcc@11.2.0~python arch=cray-sles15-zen2
                                ^xz@5.2.5%gcc@11.2.0~pic libs=shared,static arch=cray-sles15-zen2
                            ^tar@1.34%gcc@11.2.0 zip=pigz arch=cray-sles15-zen2
                                ^pigz@2.7%gcc@11.2.0 arch=cray-sles15-zen2
                                ^zstd@1.5.2%gcc@11.2.0+programs compression=none libs=shared,static arch=cray-sles15-zen2
                        ^libffi@3.4.2%gcc@11.2.0 arch=cray-sles15-zen2
                        ^sqlite@3.38.5%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=cray-sles15-zen2
                        ^util-linux-uuid@2.37.4%gcc@11.2.0 arch=cray-sles15-zen2
                    ^z3@4.8.16%gcc@11.2.0~gmp~ipo~python build_type=RelWithDebInfo arch=cray-sles15-zen2
                ^rocm-cmake@5.1.3%gcc@11.2.0~ipo build_type=Release arch=cray-sles15-zen2
            ^glx@1.4%gcc@11.2.0 arch=cray-sles15-zen2
                ^mesa@22.1.2%gcc@11.2.0+glx+llvm+opengl~opengles+osmesa~strip buildtype=release default_library=shared patches=ada85be arch=cray-sles15-zen2
                    ^bison@3.8.2%gcc@11.2.0 arch=cray-sles15-zen2
                        ^m4@1.4.19%gcc@11.2.0+sigsegv patches=9dc5fbd,bfdffa7 arch=cray-sles15-zen2
                            ^libsigsegv@2.13%gcc@11.2.0 arch=cray-sles15-zen2
                    ^flex@2.6.3%gcc@11.2.0+lex~nls arch=cray-sles15-zen2
                        ^findutils@4.9.0%gcc@11.2.0 patches=440b954 arch=cray-sles15-zen2
                    ^glproto@1.4.17%gcc@11.2.0 arch=cray-sles15-zen2
                        ^util-macros@1.19.3%gcc@11.2.0 arch=cray-sles15-zen2
                    ^libunwind@1.6.2%gcc@11.2.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs~pic+tests+weak_backtrace~xz~zlib components=none libs=shared,static arch=cray-sles15-zen2
                    ^libx11@1.7.0%gcc@11.2.0 arch=cray-sles15-zen2
                        ^inputproto@2.3.2%gcc@11.2.0 arch=cray-sles15-zen2
                        ^kbproto@1.0.7%gcc@11.2.0 arch=cray-sles15-zen2
                        ^libxcb@1.14%gcc@11.2.0 arch=cray-sles15-zen2
                            ^libpthread-stubs@0.4%gcc@11.2.0 arch=cray-sles15-zen2
                            ^libxau@1.0.8%gcc@11.2.0 arch=cray-sles15-zen2
                                ^xproto@7.0.31%gcc@11.2.0 arch=cray-sles15-zen2
                            ^libxdmcp@1.1.2%gcc@11.2.0 arch=cray-sles15-zen2
                            ^xcb-proto@1.14.1%gcc@11.2.0 arch=cray-sles15-zen2
                        ^xextproto@7.3.0%gcc@11.2.0 arch=cray-sles15-zen2
                        ^xtrans@1.3.5%gcc@11.2.0 arch=cray-sles15-zen2
                    ^libxext@1.3.3%gcc@11.2.0 arch=cray-sles15-zen2
                    ^libxt@1.1.5%gcc@11.2.0 arch=cray-sles15-zen2
                        ^libice@1.0.9%gcc@11.2.0 arch=cray-sles15-zen2
                        ^libsm@1.2.3%gcc@11.2.0 arch=cray-sles15-zen2
                    ^meson@0.62.2%gcc@11.2.0 patches=9c87472,aa6c50d arch=cray-sles15-zen2
                        ^ninja@1.11.0%gcc@11.2.0 arch=cray-sles15-zen2
                        ^py-pip@22.1.2%gcc@11.2.0 arch=cray-sles15-zen2
                        ^py-setuptools@63.0.0%gcc@11.2.0 arch=cray-sles15-zen2
                        ^py-wheel@0.37.1%gcc@11.2.0 arch=cray-sles15-zen2
                    ^py-mako@1.1.6%gcc@11.2.0 arch=cray-sles15-zen2
                        ^py-markupsafe@2.0.1%gcc@11.2.0 arch=cray-sles15-zen2
                    ^xrandr@1.5.0%gcc@11.2.0 arch=cray-sles15-zen2
                        ^libxrandr@1.5.0%gcc@11.2.0 arch=cray-sles15-zen2
                            ^libxrender@0.9.10%gcc@11.2.0 arch=cray-sles15-zen2
                                ^renderproto@0.11.1%gcc@11.2.0 arch=cray-sles15-zen2
                            ^randrproto@1.5.0%gcc@11.2.0 arch=cray-sles15-zen2
            ^hsa-rocr-dev@5.1.3%gcc@11.2.0+image~ipo+shared build_type=Release patches=71e6851 arch=cray-sles15-zen2
                ^elfutils@0.186%gcc@11.2.0~bzip2~debuginfod+nls~xz arch=cray-sles15-zen2
                ^hsakmt-roct@5.1.3%gcc@11.2.0~ipo+shared build_type=Release patches=f926273 arch=cray-sles15-zen2
                    ^libdrm@2.4.110%gcc@11.2.0~docs arch=cray-sles15-zen2
                        ^libpciaccess@0.16%gcc@11.2.0 arch=cray-sles15-zen2
                            ^libtool@2.4.7%gcc@11.2.0 arch=cray-sles15-zen2
                    ^numactl@2.0.14%gcc@11.2.0 patches=4e1d78c,62fc8a8,ff37630 arch=cray-sles15-zen2
                        ^autoconf@2.69%gcc@11.2.0 patches=35c4492,7793209,a49dd5b arch=cray-sles15-zen2
                        ^automake@1.16.5%gcc@11.2.0 arch=cray-sles15-zen2
                ^xxd-standalone@8.2.1201%gcc@11.2.0 arch=cray-sles15-zen2
            ^perl-file-which@1.22%gcc@11.2.0 arch=cray-sles15-zen2
            ^perl-uri-encode@1.1.1%gcc@11.2.0 arch=cray-sles15-zen2
                ^perl-module-build@0.4224%gcc@11.2.0 arch=cray-sles15-zen2
            ^rocminfo@5.1.3%gcc@11.2.0~ipo build_type=Release arch=cray-sles15-zen2
            ^roctracer-dev-api@5.1.3%gcc@11.2.0 arch=cray-sles15-zen2
```

</details>

Any ideas about what might be causing this?

In any case, I've pushed a change with `CMAKE_HIP_ARCHITECTURES` and removed the `+rocm`/`cmake` constraint (for now).

**Comment by davidbeckingsale** [comment]
@msimberg - we have also seen that issue, and @trws identified a fix, something like:

```
            entries.append(
                 cmake_cache_string(
                     "HIP_CLANG_INCLUDE_PATH",
                     glob.glob(
                         "{}/lib/clang/*/include".format(spec['llvm-amdgpu'].prefix)
                     )[0]
                 )
             )
```

I believe this change will be added to multiple packages in a separate PR

File: var/spack/repos/builtin/packages/umpire/package.py
Comment (msimberg) [reply var/spack/repos/builtin/packages/umpire/package.py]: I also tried building with CMake 3.21.0 and it actually seems to work so for some reason umpire is immune to that problem. Umpire does not (explicitly?) set `hipcc` as the C++ compiler which is probably why it's not affected.

**Comment by msimberg** [comment]
> @msimberg - we have also seen that issue, and @trws identified a fix, something like:
> 
> ```
>             entries.append(
>                  cmake_cache_string(
>                      "HIP_CLANG_INCLUDE_PATH",
>                      glob.glob(
>                          "{}/lib/clang/*/include".format(spec['llvm-amdgpu'].prefix)
>                      )[0]
>                  )
>              )
> ```
> 
> I believe this change will be added to multiple packages in a separate PR

This got me past configuration but it fails later not finding `std::filesystem` (which is fair because it's compiling with C++14... some detection is going wrong somewhere). Is this also known? Would you like me to open an issue for that, probably on the umpire repo?

Do you happen to know which versions of umpire/blt/etc. fail with the `HIP_CLANG_INCLUDE_PATH` error, i.e. would you know what conflict to add for that until that's patched up? Or would it be ok to go ahead with this in the meantime and add conflicts/fixes in separate PRs?

**Comment by msimberg** [comment]
@davidbeckingsale I think this is the problem with `std::filesystem`: https://github.com/LLNL/Umpire/issues/771.

**Comment by msimberg** [comment]
Fixed the style issues.

**Comment by msimberg** [comment]
Ping @davidbeckingsale? Is there something missing from here before it can be merged? Not having this currently causes conflicts in packages that want to use CMake's HIP language support which requires at least CMake 3.21 (in practice 3.22 because of bugs).

**Comment by haampie** [comment]
Using `CMAKE_HIP_ARCHITECTURES` would require a lowerbound on cmake version 3.21, right?

https://cmake.org/cmake/help/latest/variable/CMAKE_HIP_ARCHITECTURES.html says new in version 3.21

Should that then also be reflected in `depends_on("cmake@3.21:", when="+rocm")`? I'll ping @davidbeckingsale another time

**Comment by msimberg** [comment]
> Using `CMAKE_HIP_ARCHITECTURES` would require a lowerbound on cmake version 3.21, right?
> 
> https://cmake.org/cmake/help/latest/variable/CMAKE_HIP_ARCHITECTURES.html says new in version 3.21
> 

I think yes and no. Setting `CMAKE_HIP_ARCHITECTURES` doesn't hurt with CMake < 3.21, but it may lead to warnings about it being ignored. Unsure how important that is to silence.

> Should that then also be reflected in `depends_on("cmake@3.21:", when="+rocm")`? I'll ping @davidbeckingsale another time

3.21 is not generally required, it's just that with 3.21 or newer `CMAKE_HIP_ARCHITECTURES` needs to be set.

**Comment by trws** [comment]
> Using `CMAKE_HIP_ARCHITECTURES` would require a lowerbound on cmake version 3.21, right?

Not for Umpire, or generally any project that uses BLT.  We made that the variable that BLT uses to set the architecture regardless of cmake version.

**Comment by msimberg** [comment]
@haampie, @trws do you think I could go ahead and merge this? I'd like to get a final ok from @davidbeckingsale but I've had trouble getting hold of him for an ok, and the change is quite minimal. Maybe this message will summon him?

**Review by davidbeckingsale — APPROVED** [review]

**Comment by trws** [comment]
Assuming it passes, I'm ok with this merging.

**Comment by msimberg** [comment]
Thanks @davidbeckingsale for having a look!

**Comment by msimberg** [comment]
@spackbot run pipeline

**Comment by msimberg** [comment]
@spackbot run pipeline

**Comment by msimberg** [comment]
I guess `ci / all (pull_request)` being `skipped` means that auto-merge won't trigger, but is this ok to merge in this state? From what I can tell all necessary checks have passed (or been skipped) and this could be merged but I'm not 100% sure what the latest word on CI is...

**Review by alalazo — APPROVED** [review]

**Pull request merged**
